### PR TITLE
Work on networking

### PR DESCRIPTION
**Pull request by valentinbreiz**
This pull request contains:
- PCNETII Driver fixes.
- UDP Packet checksum now correctly calculated.
- Minor changes & fixes.

TCP will be added soon.
DHCP after.

### CODE REVIEW

File: source/Cosmos.System2/Network/IPv4/Address.cs [hunk 1]
```diff
@@ -67,7 +68,17 @@ public bool IsLoopbackAddress()
 
         public bool IsBroadcastAddress()
         {
-            if (address[0] == 255)
+            if ((address[0] == 255) && (address[1] == 255) && (address[2] == 255) && (address[3] == 255))
```
Comment (jp2masa) [inline source/Cosmos.System2/Network/IPv4/Address.cs]: Nit: can be written as:
```cs
return address[0] == 0xFF
    && address[1] == 0xFF
    && address[2] == 0xFF
    && address[3] == 0xFF;
```
Or expression body:
```cs
public bool IsBroadcastAddress() =>
    address[0] == 0xFF
    && address[1] == 0xFF
    && address[2] == 0xFF
    && address[3] == 0xFF;
```

File: source/Cosmos.HAL2/Drivers/PCI/Network/AMDPCNetII.cs [hunk 2]
```diff
@@ -319,6 +323,10 @@ private void ReadRawData()
                 if ((status & 0x80000000) == 0)
                 {
                     recv_size = (UInt16)(mRxDescriptor.Read32(xOffset + 0) & 0xFFF);
+
+                    if (recv_size > 64) // remove checksum
+                        recv_size -= 4;
```
[the remark this answers is not in the document]
Comment (valentinbreiz) [reply source/Cosmos.HAL2/Drivers/PCI/Network/AMDPCNetII.cs]: Curlies?

File: source/Cosmos.System2/Network/NetworkStack.cs [hunk 3]
```diff
@@ -21,40 +26,39 @@ public static void Init()
...
             // VMT Scanner issue workaround
-            Cosmos.System.Network.ARP.ARPPacket.VMTInclude();
-            //Cosmos.System.Network.IPv4.ARPPacket_Ethernet.VMTInclude();
-            Cosmos.System.Network.IPv4.ARPReply_Ethernet.VMTInclude();
-            Cosmos.System.Network.IPv4.ARPRequest_Ethernet.VMTInclude();
-            Cosmos.System.Network.IPv4.ICMPPacket.VMTInclude();
-            Cosmos.System.Network.IPv4.ICMPEchoReply.VMTInclude();
-            Cosmos.System.Network.IPv4.ICMPEchoRequest.VMTInclude();
-            Cosmos.System.Network.IPv4.UDPPacket.VMTInclude();
+            ARPPacket.VMTInclude();
```
[the remark this answers is not in the document]
Comment (valentinbreiz) [reply source/Cosmos.System2/Network/NetworkStack.cs]: I don't know I didn't try without them.

File: source/Cosmos.System2/Network/IPv4/OutgoingBuffer.cs [hunk 4]
```diff
@@ -46,32 +53,70 @@ internal static void AddPacket(IPPacket packet)
...
-            if (queue.Count < 1)
-            {
-                return;
-            }
+            int _deltaT = 0;
+            int second = 0;
 
-            //foreach (BufferEntry entry in queue)
-            for (int e = 0; e < queue.Count; e++)
+            while (!(queue.Count < 1))
```
[the remark this answers is not in the document]
Comment (valentinbreiz) [reply source/Cosmos.System2/Network/IPv4/OutgoingBuffer.cs]: Indeed!

File: source/Cosmos.System2/Network/IPv4/OutgoingBuffer.cs [hunk 5]
```diff
@@ -9,11 +9,10 @@ namespace Cosmos.System.Network.IPv4
 {
     internal static class OutgoingBuffer
     {
-        internal static Debugger mDebugger = new Debugger("System", "Networking");
 
         private class BufferEntry
         {
-            public enum EntryStatus { ADDED, ARP_SENT, ROUTE_ARP_SENT, JUST_SEND, DONE };
+            public enum EntryStatus { ADDED, ARP_SENT, ROUTE_ARP_SENT, JUST_SEND, DONE, DHCP_REQUEST };
```
[the remark this answers is not in the document]
Comment (valentinbreiz) [reply source/Cosmos.System2/Network/IPv4/OutgoingBuffer.cs]: PascalCase?

File: source/Cosmos.System2/Network/IPv4/Address.cs (hunk 1 above)
Comment (valentinbreiz) [reply source/Cosmos.System2/Network/IPv4/Address.cs]: Yes!

File: source/Cosmos.HAL2/Drivers/PCI/Network/AMDPCNetII.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jp2masa) [reply source/Cosmos.HAL2/Drivers/PCI/Network/AMDPCNetII.cs]: `{` and `}`.

File: source/Cosmos.System2/Network/IPv4/OutgoingBuffer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jp2masa) [reply source/Cosmos.System2/Network/IPv4/OutgoingBuffer.cs]: For example:
`ADDED` -> `Added`
`ARP_SENT` -> `ArpSent`
...

File: source/Cosmos.HAL2/Drivers/PCI/Network/AMDPCNetII.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (valentinbreiz) [reply source/Cosmos.HAL2/Drivers/PCI/Network/AMDPCNetII.cs]: What does it change?

File: source/Cosmos.System2/Network/IPv4/OutgoingBuffer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (valentinbreiz) [reply source/Cosmos.System2/Network/IPv4/OutgoingBuffer.cs]: I think it's better to have status like that.

File: source/Cosmos.HAL2/Drivers/PCI/Network/AMDPCNetII.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jp2masa) [reply source/Cosmos.HAL2/Drivers/PCI/Network/AMDPCNetII.cs]: It's part of the repo coding style, that's why it's a nit.

**Comment by valentinbreiz** [comment]
Can you merge it ?

**Comment by valentinbreiz** [comment]
Yes!

**Pull request merged**